Mathlib/Analysis/Calculus/Gradient/Basic.lean

```diff
@@ -0,0 +1,385 @@
+def HasGradientAtFilter (f : F → 𝕜) (f' x : F) (L : Filter F) :=
+  HasFDerivAtFilter f (toDual 𝕜 F f') x L
+
+/-- `f` has the gradient `f'` at the point `x` within the subset `s` if
+  `f x' = f x + ⟨ f', x' - x ⟩ + o (x' - x)` where `x'` converges to `x` inside `s`.-/
+def HasGradientWithinAt (f : F → 𝕜) (f' : F) (s : Set F) (x : F):=
+  HasGradientAtFilter f f' x (𝓝[s] x)
+
+/-- `f` has the gradient `f'` at the point `x` if
+  `f x' = f x + ⟨ f', x' - x ⟩ + o (x' - x)` where `x'` converges to `x`. -/
+def HasGradientAt (f : F → 𝕜) (f' x : F) :=
+  HasGradientAtFilter f f' x (𝓝 x)
+
+/-- Gradient of `f` at the point `x`, if it exists.  Zero otherwise.
+If the derivative exists (i.e., `∃ f', HasGradientAt f f' x`), then
+`f x' = f x + ⟨ f', x' - x ⟩ + o (x' - x)` where `x'` converges to `x`. -/
+def Gradient (f : F → 𝕜) (x : F) : F := (toDual 𝕜 F).symm (fderiv 𝕜 f x)
+
+local notation "∇" => Gradient
+
+variable {s : Set F} {L : Filter F}
+
+theorem HasGradientWithinAt_iff_HasFDerivWithinAt {s : Set F} :
+  HasGradientWithinAt f f' s x ↔ HasFDerivWithinAt f (toDual 𝕜 F f') s x := by
+  constructor <;> (intro h; exact h)
+
+theorem HasGradientWithinAt.hasFDerivWithinAt {s : Set F} (h : HasGradientWithinAt f f' s x) :
+  HasFDerivWithinAt f (toDual 𝕜 F f') s x := HasGradientWithinAt_iff_HasFDerivWithinAt.mp h
+
+theorem HasFDerivWithinAt.hasGradientWithinAt {s : Set F}
+  (h : HasFDerivWithinAt f (toDual 𝕜 F f') s x) :
+  HasGradientWithinAt f f' s x := HasGradientWithinAt_iff_HasFDerivWithinAt.mpr h
+
+theorem HasGradientAt_iff_HasFDerivAt :
+  HasGradientAt f f' x ↔ HasFDerivAt f (toDual 𝕜 F f') x := by
+  constructor <;> (intro h; exact h)
+
+theorem HasGradientAt.hasFDerivAt (h : HasGradientAt f f' x) :
+  HasFDerivAt f (toDual 𝕜 F f') x := HasGradientAt_iff_HasFDerivAt.mp h
+
+theorem HasFDerivAt.hasGradientAt (h : HasFDerivAt f (toDual 𝕜 F f') x) :
+  HasGradientAt f f' x := HasGradientAt_iff_HasFDerivAt.mpr h
+
+theorem Gradient_zero_of_not_differentiableAt (h : ¬DifferentiableAt 𝕜 f x) : ∇ f x = 0 := by
+  have : (toDual 𝕜 F).symm 0 = 0 := by simp only [map_zero]
+  rw [← fderiv_zero_of_not_differentiableAt h] at this
+  exact this
+
+local notation "⟪" x ", " y "⟫" => @inner 𝕜 _ _ x y
+
+theorem toDual_eq_inner (y x : F) : ((toDual 𝕜 F) y) x = ⟪y, x⟫ := rfl
+
+theorem toDual_of_toDual_frechet (frechet : F →L[𝕜] 𝕜) :
+  frechet = (toDual 𝕜 F) ((toDual 𝕜 F).symm frechet) := by
+  apply ContinuousLinearMap.ext_iff.mpr
+  intro x
+  have : ((toDual 𝕜 F) ((toDual 𝕜 F).symm frechet)) x = ⟪(toDual 𝕜 F).symm frechet, x⟫ := rfl
+  rw [this]
+  apply toDual_symm_apply.symm
+
+theorem toDual_of_toDual_gradient (gradient : F) :
+  gradient = (toDual 𝕜 F).symm ((toDual 𝕜 F) gradient) :=
+   Eq.symm (LinearIsometryEquiv.symm_apply_apply (toDual 𝕜 F) gradient)
+
+section GradientUniqueness
+
+variable {gradf gradg : F}
+
+theorem toDual_unique {f₁ g₁ : F} (h : ((toDual 𝕜 F) f₁) = ((toDual 𝕜 F) g₁)) : f₁ = g₁ := by
```

Comment:
`toDual 𝕜 F` is an equiv, so it is injective and sujective by general properties: the proof of the theorem should be a one-liner but, more importantly, you shouldn't need the theorem at all!

Reply:
Thank you for your advice. We totally agree that there is no need to have this theorem. So we update our proof.